subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyCollector.java

```diff
@@ -17,33 +17,44 @@
 package org.gradle.api.internal.artifacts.dsl.dependencies;
 
 import org.gradle.api.Action;
+import org.gradle.api.DomainObjectSet;
 import org.gradle.api.InvalidUserCodeException;
-import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.artifacts.ExternalModuleDependency;
 import org.gradle.api.artifacts.FileCollectionDependency;
 import org.gradle.api.artifacts.MinimalExternalModuleDependency;
-import org.gradle.api.artifacts.dsl.DependencyAdder;
+import org.gradle.api.artifacts.dsl.DependencyCollector;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.provider.ProviderConvertible;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.tasks.Nested;
 
 import javax.annotation.Nullable;
 import javax.inject.Inject;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.Set;
 
-public class DefaultDependencyAdder implements DependencyAdder {
+public abstract class DefaultDependencyCollector implements DependencyCollector {
     private final DependencyFactoryInternal dependencyFactory;
-    private final Configuration configuration;
+    private final ProviderFactory providerFactory;
 
     @Inject
-    public DefaultDependencyAdder(DependencyFactoryInternal dependencyFactory, Configuration configuration) {
+    public DefaultDependencyCollector(DependencyFactoryInternal dependencyFactory, ProviderFactory providerFactory) {
         this.dependencyFactory = dependencyFactory;
-        this.configuration = configuration;
+        this.providerFactory = providerFactory;
     }
 
+    @Override
+    public Provider<Set<Dependency>> getDependencies() {
+        return providerFactory.provider(this::getDeclaredDependencies);
+    }
+
+    @Nested
+    protected abstract DomainObjectSet<Dependency> getDeclaredDependencies();
```

Comment:
🤔 Why not use a `SetProperty<Dependency>` and just return it as the `Provider`, instead of separating them by a callable `Provider`?

Reply:
Pushed up an update